### ISSUE TITLE
[enhancement]: Backwards Compatibility

### ISSUE THREAD

**Issue by Neosettler** [invoke-ai/InvokeAI#2459]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What should this feature add?

Greetings, 

It’s been a blast generating images with InvokeAI until… I find out that newer version were not able to render images generated with previous versions. Sometimes with subtle changes, sometimes completely unrecognizable. Digging deeper, @lstein find out that the some optimization has been done that completely broke backward compatibility.

It is assumed that a simple prompt like “red cat” with the same parameters would give the same result in any fork of stable-diffusion, correct? If not, that’s terribly wrong!

Basically, it should always be possible to go back to a vanilla state and if optimizations are applied, it should be clearly stated and optional. (This should be true for any tempering of the rendering)

All that to say that, if the migration to Diffuser that is currently cooking has to start from scratch, please consider making backward compatibility one of the development pillars.

Thank you,

**Comment by Barafu** [comment]
This is a very advanced request. You can always download older version of InvokeAI into a separate folder and run it. Maintaining the host of outdated versions of algorithms for such a rare need is too complex. The devs are clearly overworked without it. 

If you said "old versions of InvokeAI should remain available", I'd totally agree. And I think they are available. But, besides InvokeAI, there is also a problem of models from HuggingFace. And pip dependencies, I think at least some of them are not version constrained and can affect generation. If you want to identically recreate images in the future, better save everything you used now. 

Why do you even need this? Maybe we can suggest a better option?

**Comment by keturn** [comment]
The "red cat" example is simple enough that I expect you _can_ reproduce it.

But this whole technology is only months old, we're learning as we go, and trying things out along the way. As we learn more about which approaches are efficient and useful, and other related technologies come out that we want to integrate, we'll need to adapt accordingly. Not every idea we had in the very first weeks and months of this project will be worth sticking to in the long term.

So while I expect "red cat" will keep doing what "red cat" does (with the same model and a predictable scheduler), "placid (red- cat)1.41 on a (blanket+).swap(aeroplane--) [cougar] [zebra+++] --threshold 4 --perlin 0.3 --hires_fix --variation_amount 0.075 --embiggen --slartibartfast" might not.

> And pip dependencies, I think at least some of them are not version constrained and can affect generation. 

I think having reproducible _releases_ is a goal our distribution & installation contributors are probably on board for. Application packaging is a complicated thing and ours has been in a lot of flux lately, so I'm not sure which releases have had frozen dependencies, but in general I think the ability to download and install previous releases is an achievable goal.

**Comment by Neosettler** [comment]
Thank you for your input @Barafu, 

There is several ways to answer your question but In a nut shell, generating more images than one can chew at the moment of exploration implies that ideas could, and probably will be revisited at a later time. To state the obvious, the digital world would be quite different if backward compatibility wasn’t well understood and/or didn’t exist as a development pillar. Point in case, I’ve choose to sink in a lot of efforts and confidence in InvokeAI and I’d feel much more comfortable knowing that mechanics and/or code paradigm are put in place to support backward compatibility going forward. Or at least, being open to the idea would be a good start.

I’m still digging but I received few commentaries about this topic from Automatic1111’s community, like this one:
“Just an FYI, Automatic1111 actively keeps generation backwards compatible. Even if a small change to sampler settings improves only small details, you can still turn it off and have the original implementation.”

This is the default behavior most user would expect quite frankly.

@keturn dependencies are pretty much the easiest part to solve. Most software that I know of that uses Python ship with prepackage solution. There is nothing more to it... hence, forcing the pip shenanigan on the user sucks, period!

**Comment by Neosettler** [comment]
Closing has clearly not happening.
